onmt/translate/beam_search.py

```diff
@@ -336,6 +238,170 @@ def update_finished(self):
+                 stepwise_penalty, ratio):
+        super(BeamSearch, self).__init__(beam_size, batch_size, pad, bos, eos,
+                                         n_best, global_scorer, min_length,
+                                         max_length, return_attention,
+                                         block_ngram_repeat, exclusion_tokens,
+                                         stepwise_penalty, ratio)
+
+    def initialize(self, memory_bank, src_lengths, src_map=None, device=None,
+                   target_prefix=None):
+        """Initialize for decoding.
+        Repeat src objects `beam_size` times.
+        """
+
+        def fn_map_state(state, dim):
+            return tile(state, self.beam_size, dim=dim)
+
+        if isinstance(memory_bank, tuple):
+            memory_bank = tuple(tile(x, self.beam_size, dim=1)
+                                for x in memory_bank)
+            mb_device = memory_bank[0].device
+        else:
+            memory_bank = tile(memory_bank, self.beam_size, dim=1)
+            mb_device = memory_bank.device
+        if src_map is not None:
+            src_map = tile(src_map, self.beam_size, dim=1)
+        if device is None:
+            device = mb_device
+
+        self.memory_lengths = tile(src_lengths, self.beam_size)
+        if target_prefix is not None:
+            target_prefix = tile(target_prefix, self.beam_size, dim=1)
+
+        super(BeamSearch, self).initialize_(
+            memory_bank, self.memory_lengths, src_map, device, target_prefix)
+
+        return fn_map_state, memory_bank, self.memory_lengths, src_map
+
+
+class BeamSearchGenerate(BeamSearchBase):
+    """
+        Beam search for language/decoder only models
+    """
+    def initialize(self, src, src_lengths, src_map=None, device=None,
+                   target_prefix=None):
+        """Initialize for decoding.
+        Repeat src objects `beam_size` times.
+        """
+        def fn_map_state(state, dim):
+            return tile(state, self.beam_size, dim=dim)
+
+        src = fn_map_state(src, dim=1)
+        if src_map is not None:
+            src_map = tile(src_map, self.beam_size, dim=1)
+        if device is None:
+            device = src.device
+
+        self.memory_lengths = tile(src_lengths, self.beam_size)
+        if target_prefix is not None:
+            target_prefix = tile(target_prefix, self.beam_size, dim=1)
+
+        super(BeamSearchGenerate, self).initialize_(
+            None, self.memory_lengths, src_map=src_map, device=device,
+            target_prefix=target_prefix)
+
+        return fn_map_state, src, self.memory_lengths, src_map
+
+    def remove_finished_batches(self, _B_new, _B_old, non_finished,
+                                predictions, attention, step):
+        super(BeamSearchGenerate, self).remove_finished_batches(
+            _B_new, _B_old, non_finished, predictions, attention, step)
+        self.memory_lengths = self.memory_lengths.view(
```

Comment:
Maybe a comment wouldn't hurt here, to help grasp the difference with the base method.